api/controllers/index.go

```diff
@@ -30,28 +32,46 @@ func IndexDiff(rw http.ResponseWriter, r *http.Request) *httperr.Error {
 	return RenderJson(rw, missing)
 }
 
-func IndexUpload(rw http.ResponseWriter, r *http.Request) *httperr.Error {
-	hash := mux.Vars(r)["hash"]
-
-	file, _, err := r.FormFile("data")
+// IndexUpdate accepts a tarball of changes to the index
+func IndexUpdate(rw http.ResponseWriter, r *http.Request) *httperr.Error {
+	update, _, err := r.FormFile("update")
 	if err != nil {
 		return httperr.Server(err)
 	}
 
-	data, err := ioutil.ReadAll(file)
+	gz, err := gzip.NewReader(update)
 	if err != nil {
 		return httperr.Server(err)
 	}
 
-	sum := sha256.Sum256(data)
+	tr := tar.NewReader(gz)
 
-	if hash != hex.EncodeToString(sum[:]) {
-		return httperr.New(403, fmt.Errorf("invalid hash"))
-	}
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return httperr.Server(err)
+		}
 
-	err = provider.IndexUpload(hash, data)
-	if err != nil {
-		return httperr.Server(err)
+		fmt.Printf("header = %+v\n", header)
```

Comment:
This printed out a lot logs for the rails example. Don't think we really need it.